packages/backend-ai/src/conversation/factory.test.ts

```diff
@@ -0,0 +1,125 @@
+import {
+  ConstructFactoryGetInstanceProps,
+  ResourceNameValidator,
+} from '@aws-amplify/plugin-types';
+import assert from 'node:assert';
+import { StackMetadataBackendOutputStorageStrategy } from '@aws-amplify/backend-output-storage';
+import {
+  ConstructContainerStub,
+  ResourceNameValidatorStub,
+  StackResolverStub,
+} from '@aws-amplify/backend-platform-test-stubs';
+import { defaultEntryHandler } from './test-assets/with-default-entry/resource.js';
+import { customEntryHandler } from './test-assets/with-custom-entry/resource.js';
+import { Template } from 'aws-cdk-lib/assertions';
+import { defineConversationHandlerFunction } from './factory.js';
+
+const createStackAndSetContext = (): Stack => {
+  const app = new App();
+  app.node.setContext('amplify-backend-name', 'testEnvName');
+  app.node.setContext('amplify-backend-namespace', 'testBackendId');
+  app.node.setContext('amplify-backend-type', 'branch');
+  const stack = new Stack(app);
+  return stack;
+};
+
+void describe('ConversationHandlerFactory', () => {
+  let rootStack: Stack;
+  let getInstanceProps: ConstructFactoryGetInstanceProps;
+  let resourceNameValidator: ResourceNameValidator;
+
+  beforeEach(() => {
+    rootStack = createStackAndSetContext();
+
+    const constructContainer = new ConstructContainerStub(
+      new StackResolverStub(rootStack)
+    );
+
+    const outputStorageStrategy = new StackMetadataBackendOutputStorageStrategy(
+      rootStack
+    );
+
+    resourceNameValidator = new ResourceNameValidatorStub();
+
+    getInstanceProps = {
+      constructContainer,
+      outputStorageStrategy,
+      resourceNameValidator,
+    };
+  });
+
+  void it('creates singleton instance', () => {
+    const factory = defaultEntryHandler;
+    const instance1 = factory.getInstance(getInstanceProps);
+    const instance2 = factory.getInstance(getInstanceProps);
+    assert.strictEqual(instance1, instance2);
+  });
+
+  void it('resolves default entry when not specified', () => {
+    const factory = defaultEntryHandler;
+    const lambda = factory.getInstance(getInstanceProps);
+    const template = Template.fromStack(Stack.of(lambda.resources.lambda));
+    template.resourceCountIs('AWS::Lambda::Function', 1);
+    template.hasResourceProperties('AWS::Lambda::Function', {
+      Handler: 'index.handler',
+    });
+  });
+
+  void it('uses entry that is explicitly specified', () => {
+    const factory = customEntryHandler;
+    const lambda = factory.getInstance(getInstanceProps);
+    const template = Template.fromStack(Stack.of(lambda.resources.lambda));
+    template.resourceCountIs('AWS::Lambda::Function', 1);
+    template.hasResourceProperties('AWS::Lambda::Function', {
+      Handler: 'index.handler',
+    });
+  });
+
+  void it('throws on invalid name', () => {
+    mock
+      .method(resourceNameValidator, 'validate')
+      .mock.mockImplementationOnce(() => {
+        throw new Error('test validation error');
```

Comment:
why mock this if this is what you're testing? we should just be looking for the error containing ```Resource name contains invalid characters, found ```

Reply:
I borrowed this from functions.

It seems to be written there this way because the unit under test is `factory` not `resourceNameValidator`. Which is correct approach if the goal is "true unit test". `resourceNameValidator` has it's own unit tests tests.

We could use real `resourceNameValidator` give it's simplicity, but we'd make this test less "unit-ish".

Reply:
nit: test name should be more clear like "throws if resourceNameValidator detects an invalid name"